core/src/Kokkos_CopyViews.hpp

```diff
@@ -3564,6 +3564,54 @@ auto create_mirror(Kokkos::Impl::WithoutInitializing_t wi, Space const&,
 
 namespace Impl {
 
+template <class T, class... P, class... ViewCtorArgs>
+inline auto choose_create_mirror(
+    const Kokkos::View<T, P...>& src,
+    const Impl::ViewCtorProp<ViewCtorArgs...>& arg_prop) {
+  // Due to the fact that users can overload `Kokkos::create_mirror`, but also
+  // that they may not have implemented all of its different possible
+  // variations, this function chooses the correct private or public version of
+  // it to call.
+
+  if constexpr (std::is_void_v<typename ViewTraits<T, P...>::specialize>) {
+    // if the view is not specialized, just call the Impl function
+    return Kokkos::Impl::create_mirror(src, arg_prop);
+  } else {
+    // otherwise, recreate the public call
+    using ViewProp = Impl::ViewCtorProp<ViewCtorArgs...>;
+    if constexpr (sizeof...(ViewCtorArgs) == 0) {
+      // if there are no view constructor args, call the specific public
+      // function
+      return Kokkos::create_mirror(src);
+    } else if constexpr (sizeof...(ViewCtorArgs) == 1 &&
+                         ViewProp::has_memory_space) {
+      // if there is one view constructor arg and it has a memory space, call
+      // the specific public function
+      return Kokkos::create_mirror(typename ViewProp::memory_space{}, src);
+    } else if constexpr (sizeof...(ViewCtorArgs) == 1 &&
+                         !ViewProp::initialize) {
+      // if there is one view constructor arg and it has a without initializing
+      // mark, call the specific public function
+      return Kokkos::create_mirror(
+          typename Kokkos::Impl::WithoutInitializing_t{}, src);
+    } else if constexpr (sizeof...(ViewCtorArgs) == 2 &&
+                         ViewProp::has_memory_space && !ViewProp::initialize) {
+      // if there is two view constructor args and they have a memory space and
+      // a without initializing mark, call the specific public function
+      return Kokkos::create_mirror(
+          typename Kokkos::Impl::WithoutInitializing_t{},
+          typename ViewProp::memory_space{}, src);
+    } else {
+      // if there are other constructor args, call the generic public function
+      return Kokkos::create_mirror(arg_prop, src);
```

Comment:
Ah so the reason that we don't call this always is that Sacado has the overload which say takes "WithoutInitializing_t" but not a general ViewCtorProp?

Reply:
That is also our understanding of the issue. It seems there are multiple overloads of the `create_mirror` function https://github.com/trilinos/Trilinos/blob/d71aa67247678321da0b4ca2e822f28b8aed8db4/packages/sacado/src/Kokkos_View_Fad_Fwd.hpp#L102 but no overload with `ViewCtorProp`.

Reply:
Yes, `Sacado` didn't follow suit when we introduced the `CiewCtorProp` overloads.

Reply:
Could we come up with a rule about this set of overloads ? For example if a user defines `Kokkos::create_mirror(Kokkos::Impl::WithoutInitializing_t, View)` then it must also define `Kokkos::create_mirror(Kokkos::Impl::ViewCtorProp<Kokkos::Impl::WithoutInitializing_t> const&, View)`

Reply:
How likely would it be for Sacado to implement the missing `create_mirror`?

Reply:
> How likely would it be for Sacado to implement the missing create_mirror?

I don't think that will happen if it's not absolutely necessary.

Reply:
So Kokkos must not assume any `specialize` overload right ? Can a user overload `Kokkos::create_mirror(Kokkos::Impl::ViewCtorProp<Kokkos::Impl::WithoutInitializing_t> const&, View)` but not `Kokkos::create_mirror(Kokkos::Impl::WithoutInitializing_t, View)` ?

Reply:
None of that is formalized anywhere and Sacado is the only one using `specialize` to the best of my knowledge.

Reply:
Greping Trilinos, it seems that Stokhos also relies on this feature.

Reply:
Yeah, `Stokhos` depends on `Sacado` and I tend to refer to them as the same. 🙂